src/Shared/LogBuffering/SerializedLogRecordFactory.cs

```diff
@@ -0,0 +1,72 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Collections.Generic;
+using System.Runtime.CompilerServices;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.ObjectPool;
+using Microsoft.Shared.Pools;
+
+namespace Microsoft.Extensions.Diagnostics.Buffering;
+
+internal static class SerializedLogRecordFactory
+{
+    private static readonly ObjectPool<List<KeyValuePair<string, object?>>> _attributesPool =
+        PoolFactory.CreateListPool<KeyValuePair<string, object?>>();
+
+    private static readonly int _serializedLogRecordSize = Unsafe.SizeOf<SerializedLogRecord>();
+
+    public static SerializedLogRecord Create(
+        LogLevel logLevel,
+        EventId eventId,
+        DateTimeOffset timestamp,
+        IReadOnlyList<KeyValuePair<string, object?>> attributes,
+        Exception? exception,
+        string formattedMessage)
+    {
+        int sizeInBytes = _serializedLogRecordSize;
+        List<KeyValuePair<string, object?>> serializedAttributes = _attributesPool.Get();
+        for (int i = 0; i < attributes.Count; i++)
+        {
+            string key = attributes[i].Key;
+            string value = attributes[i].Value?.ToString() ?? string.Empty;
+
+            sizeInBytes += key.Length * sizeof(char);
```

Comment:
Hmmm. Although this is technically mostly correct, it is annoying that we're counting the space for something that will typically be a singleton in memory (the string "Region" for example). Maybe we should just skip the key, given that it will almost always be a string constant?

Additionally, for cases in this code that measure the # of bytes in a string, you should include another N bytes (approximately) for object headers and such. I don't know what Unsafe.SizeOf<String> would return, but maybe that'd be the right thing to use.